include/z64camera.h

```diff
@@ -27,7 +28,89 @@
 #define PARENT_CAM(cam) ((cam)->play->cameraPtrs[(cam)->parentCamId])
 #define CHILD_CAM(cam) ((cam)->play->cameraPtrs[(cam)->childCamId])
 
-// All scenes using `SCENE_CAM_TYPE_FIXED_SHOP_VIEWPOINT` or `SCENE_CAM_TYPE_FIXED_TOGGLE_VIEWPOINT` are expected
+#define CAM_HUD_VISIBILITY_SHIFT 8
+#define CAM_HUD_VISIBILITY_MASK (0x0F00)
+#define CAM_HUD_VISIBILITY(hudVisibility) (((hudVisibility) & 0xF) << CAM_HUD_VISIBILITY_SHIFT)
+
+#define CAM_HUD_VISIBILITY_ALL CAM_HUD_VISIBILITY(0) // HUD_VISIBILITY_ALL
+#define CAM_HUD_VISIBILITY_NOTHING CAM_HUD_VISIBILITY(HUD_VISIBILITY_NOTHING)
+#define CAM_HUD_VISIBILITY_NOTHING_ALT CAM_HUD_VISIBILITY(HUD_VISIBILITY_NOTHING_ALT)
+#define CAM_HUD_VISIBILITY_HEARTS_FORCE CAM_HUD_VISIBILITY(HUD_VISIBILITY_HEARTS_FORCE)
+#define CAM_HUD_VISIBILITY_A CAM_HUD_VISIBILITY(HUD_VISIBILITY_A)
+#define CAM_HUD_VISIBILITY_A_HEARTS_MAGIC_FORCE CAM_HUD_VISIBILITY(HUD_VISIBILITY_A_HEARTS_MAGIC_FORCE)
+#define CAM_HUD_VISIBILITY_A_HEARTS_MAGIC_MINIMAP_FORCE CAM_HUD_VISIBILITY(HUD_VISIBILITY_A_HEARTS_MAGIC_MINIMAP_FORCE)
+#define CAM_HUD_VISIBILITY_ALL_NO_MINIMAP_BY_BTN_STATUS CAM_HUD_VISIBILITY(HUD_VISIBILITY_ALL_NO_MINIMAP_BY_BTN_STATUS)
+#define CAM_HUD_VISIBILITY_B CAM_HUD_VISIBILITY(HUD_VISIBILITY_B)
+#define CAM_HUD_VISIBILITY_HEARTS_MAGIC CAM_HUD_VISIBILITY(HUD_VISIBILITY_HEARTS_MAGIC)
+#define CAM_HUD_VISIBILITY_B_ALT CAM_HUD_VISIBILITY(HUD_VISIBILITY_B_ALT)
+#define CAM_HUD_VISIBILITY_HEARTS CAM_HUD_VISIBILITY(HUD_VISIBILITY_HEARTS)
+#define CAM_HUD_VISIBILITY_A_B_MINIMAP CAM_HUD_VISIBILITY(HUD_VISIBILITY_A_B_MINIMAP)
+#define CAM_HUD_VISIBILITY_HEARTS_MAGIC_FORCE CAM_HUD_VISIBILITY(HUD_VISIBILITY_HEARTS_MAGIC_FORCE)
```

Comment:
I'm not opposed to using `HUD_VISIBILITY_` directly, but there might need to be another case for `HUD_VISIBILITY_NONE_INSTANT` since using that would result in unexpected behavior (would result in `HUD_VISIBILITY_A` after being masked by `& 0xF`).

but there's 1 other reason why I chose to make a separate `CAM_HUD_VISIBILITY_` enum, although this reason only applies to MM, so I'd understand if it was less of a concern here in OoT. I'd like like your thoughts on this though:

In OoT, the core of the hudVisibility enum ranges from `0x0 - 0xD`, so there's no issues only using `& 0xF` from the camera bits/field for interface params, which the exceptions of values `50` and `52`

In MM however, the core of the hudVisibility enum ranges from `0x0 - 0x16` which extends past the `& 0xF` range given in the camera interface bits/field, so all these values would not be usable directly in the enum, and would result in the wrong hudVisibility being set:
https://github.com/zeldaret/mm/blob/90469fd44220b0d40c1b7d0091089867cc47ef34/include/z64save.h#L131-L137

In the end, this could just all be explained by comments, but then it's still not the safest macro

Reply:
```
#define CAM_HUD_VISIBILITY(hudVisibilityMode) \
    (((hudVisibilityMode) == HUD_VISIBILITY_ALL ? 0 : (hudVisibilityMode) >= 0x10 ? 0xF : ((hudVisibilityMode) & 0xF)) << CAM_HUD_VISIBILITY_SHIFT)
```
Hmmm, could possibly catch values `0x10-0x16` and `0x52` in a single check

Reply:
In OoT, would also need to account for `/* 0 */ HUD_VISIBILITY_NO_CHANGE,` being used, because that would also lead to unexpected behavior. One would expect "no change" but instead would get the behavior of `HUD_VISIBILITY_ALL`

Reply:
```
#define CAM_HUD_VISIBILITY(hudVisibilityMode) \
    (((hudVisibilityMode) == HUD_VISIBILITY_ALL ? 0 : \
    ((hudVisibilityMode) == HUD_VISIBILITY_NO_CHANGE || (hudVisibilityMode) >= 0xF) ? (0xF << CAM_HUD_VISIBILITY_SHIFT) : \
    ((hudVisibilityMode) & 0xF)) << CAM_HUD_VISIBILITY_SHIFT)
```
:Thonk:

Reply:
Hmm yeah I have to agree it wouldn't be ideal to have values that are outright not supported, and as you said we could default to `IGNORE` but that's just going lead to unexpected behavior (especially for values 0xF-0x16 in MM, but this is true even for `HUD_VISIBILITY_NOTHING_INSTANT` in OoT). I was going to suggest we include some form of static assert in the macro to generate a compiler error when it's used with unsupported values, but I don't think that could work due to how the macro is used, especially inside code.

Reply:
So yeah I guess we can stay with the duplicated subset of defines, this should help in consistency with MM and it arguably keeps things "simpler", in a sense. Tho it might be worth leaving a comment to mention that only a specific range of values is supported by the format and that's why the `HUD_VISIBILITY_*` enum isn't used directly (and it could help in case anyone tries to go through this whole process again 😅).

Thanks for the detailed answer btw, I wasn't aware of the MM case (I actually assumed it would be better to deal with in MM and that they refactored it into something simpler), and I totally missed the `HUD_VISIBILITY_NO_CHANGE` => `HUD_VISIBILITY_ALL` issue 👍

Reply:
So to summarize the challenge:

For OoT: the 3 special values that don't map 1-to-1 are:
`HUD_VISIBILITY_ALL = 50` which maps to `0` for cam-hudVisibility
`HUD_VISIBILITY_NOTHING_INSTANT = 52` which has no mapping for cam hudVisibility (interpreted as `HUD_VISIBILITY_A`)
`HUD_VISIBILITY_NO_CHANGE = 0` which has no mapping for cam hudVisibility (interpreted as `HUD_VISIBILITY_ALL`)

And then there's a cam hudVisibility of `0xF` which has no mapping to a hudVisibility (but is roughly the same behavior as `HUD_VISIBILITY_NO_CHANGE`) by skipping over the camera code of processing cam-hudVisibility

In MM, there's all the above plus enum values `0xF-0x16` which all have no mapping for cam hudVisibility, although that seems more of a result of just hacking on more cases after-the-fact without taking cam-hudVisibility into account (+1 for  MM being just an OoT RomHack)

Reply:
Also I suggest aligning all the CAM_HUD_VISIBILITY_ defines vertically so it's a bit more readable and easier to check which hud visibility modes are supported

Reply:
Oh, beat me by a few seconds. Yeah, so while the `CAM_HUD_VISIBILITY_` enums are tedious and annoying to have a second set of the same thing, at least it keeps the usage known to the user and they don't have to worry about all these special cases I listed above